File: frontend/src/types/statistics.ts
/** Types mirroring `backend/schemas/statistics.py`. */

export interface MetricSummary {
  avg: number | null;
  median: number | null;
  min: number | null;
  max: number | null;
}

export interface Overview {
  count: number;
  price: MetricSummary;
  total_monthly_cost: MetricSummary;
}

export interface DistrictStats {
  district: string;
  count: number;
  price: MetricSummary;
  total_monthly_cost: MetricSummary;
}

export interface PriceHistogramBucket {
  from: number;
  to: number;
  count: number;
}

export interface DepositsStats {
  with_deposit: number;
  without_deposit: number;
  unknown: number;
  with_deposit_percent: number | null;
  avg_deposit: number | null;
  median_deposit: number | null;
}

export interface AdditionalCostsStats {
  with_additional_cost: number;
  without_additional_cost: number;
  unknown: number;
  with_additional_cost_percent: number | null;
  avg_additional_cost: number | null;
  median_additional_cost: number | null;
}

export interface NegotiationStats {
  negotiable: number;
  non_negotiable: number;
  negotiable_percent: number | null;
}

File: backend/schemas/statistics.py
from __future__ import annotations

from typing import Optional

from pydantic import BaseModel


class MetricSummary(BaseModel):
    avg: Optional[float] = None
    median: Optional[float] = None
    min: Optional[float] = None
    max: Optional[float] = None


class OverviewOut(BaseModel):
    count: int
    price: MetricSummary
    total_monthly_cost: MetricSummary


class DistrictStatsOut(BaseModel):
    district: str
    count: int
    price: MetricSummary
    total_monthly_cost: MetricSummary


class DepositsStatsOut(BaseModel):
    with_deposit: int
    without_deposit: int
    unknown: int
    with_deposit_percent: Optional[float] = None
    avg_deposit: Optional[float] = None
    median_deposit: Optional[float] = None


class AdditionalCostsStatsOut(BaseModel):
    with_additional_cost: int
    without_additional_cost: int
    unknown: int
    with_additional_cost_percent: Optional[float] = None
    avg_additional_cost: Optional[float] = None
    median_additional_cost: Optional[float] = None


class NegotiationStatsOut(BaseModel):
    negotiable: int
    non_negotiable: int
    negotiable_percent: Optional[float] = None
